tests/test_bolt2-01-close_channel.py

```diff
@@ -0,0 +1,187 @@
+     ______________________________________________________
+    | runner -> shutdown (wrong script pub key) -> ln-node |
+    | runner <-         warning msg             <- ln-node |
+    -------------------------------------------------------
+    """
+    namespaceoverride(peer_message_namespace())
+
+    # the option that the helper method feels for us
+    test_opts = {}
+    pre_events_conn = connect_to_node_helper(
+        runner, tx_spendable=tx_spendable, conn_privkey="03"
+    )
+    pre_events = open_and_announce_channel_helper(
+        runner, conn_privkey="03", opts=test_opts
+    )
+    # merge the two events
+    pre_events = merge_events_sequences(pre_events_conn, pre_events)
+    channel_idx = channel_id()
+
+    script = BitcoinUtils.build_valid_script(ScriptType.INVALID_CLOSE_SCRIPT)
+
+    test = [
+        # runner sent shutdown message to the ln implementation
+        TryAll(
+            [
+                Msg(
+                    "shutdown",
+                    channel_id=channel_idx,
+                    scriptpubkey=script,
+                ),
+                MustNotMsg("shutdown"),
+                MustNotMsg("add_htlc"),
+                # FIXME: add support for warning messages in pyln package?
+            ],
+        ),
+    ]
+    run_runner(runner, merge_events_sequences(pre=pre_events, post=test))
+
+
+def test_close_channel_allow_multiple_shutdown_msg_receive_side(
+    runner: Runner, namespaceoverride: Any
+) -> None:
+    """
+    Close operation from the receiver point of view, where the receiver will
+    receive multiple shutdown msg.
+
+    FIXME: this need to be allowed by the spec!
+     ________________________________
+    | runner -> shutdown -> ln-node |
+    | runner -> shutdown -> ln-node |
+    | runner <- shutdown <- ln-node |
+    --------------------------------
+    """
+    namespaceoverride(peer_message_namespace())
+    test_opts = {}
+    pre_events_conn = connect_to_node_helper(
+        runner, tx_spendable=tx_spendable, conn_privkey="03"
+    )
+    pre_events = open_and_announce_channel_helper(
+        runner, conn_privkey="03", opts=test_opts
+    )
+    # merge the two events
+    pre_events = merge_events_sequences(pre_events_conn, pre_events)
+    channel_idx = channel_id()
+
+    script = BitcoinUtils.build_valid_script()
+
+    test = [
+        # runner sent shutdown message to the ln implementation
+        TryAll(
+            [
+                Msg(
+                    "shutdown",
+                    channel_id=channel_idx,
+                    scriptpubkey=script,
+                ),
+                Msg(
+                    "shutdown",
+                    channel_id=channel_idx,
+                    scriptpubkey=script,
+                ),
+                MustNotMsg("update_add_htlc"),
+                # TODO: must be -> ExpectMsg("shutdown", channel_id=channel_idx, scriptpubkey=script),
+                # why the script is different
```

Comment:
@vincenzopalazzo  what do you mean by this? Do you mean that BitcoinUtils.build_valid_script() should return the same script everytime?

Reply:
In the BOLT2 there is the following assertion

MUST set scriptpubkey in one of the following forms:

OP_0 20 20-bytes (version 0 pay to witness pubkey hash), OR
OP_0 32 32-bytes (version 0 pay to witness script hash), OR
if (and only if) option_shutdown_anysegwit is negotiated:
OP_1 through OP_16 inclusive, followed by a single push of 2 to 40 bytes (witness program versions 1 through 16)

Maybe the name is not self explicative? better `build_valid_ln_script`?